python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,280 @@
+                       QgsFeatureSink,
+                       QgsProcessingAlgorithm,
+                       QgsProcessingParameterFeatureSource,
+                       QgsProcessingParameterFeatureSink,
+                       QgsProcessingOutputNumber,
+                       QgsWkbTypes,
+                       QgsFields,
+                       QgsField)
+
+from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
+
+
+class Climb(QgisAlgorithm):
+
+    INPUT = 'INPUT'
+    OUTPUT = 'OUTPUT'
+
+    TOTALCLIMB = 'TOTALCLIMB'
+    TOTALDESCENT = 'TOTALDESCENT'
+    MINELEVATION = 'MINELEVATION'
+    MAXELEVATION = 'MAXELEVATION'
+    CLIMBATTRIBUTE = 'climb'
+    DESCENTATTRIBUTE = 'descent'
+    MINELEVATTRIBUTE = 'minelev'
+    MAXELEVATTRIBUTE = 'maxelev'
+
+    def name(self):
+        return 'climbalongline'
+
+    def displayName(self):
+        return self.tr('Climb Along Line')
+
+    def group(self):
+        return self.tr('Vector analysis')
+
+    def groupId(self):
+        return 'vectoranalysis'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(
+            QgsProcessingParameterFeatureSource(
+                self.INPUT,
+                self.tr('Input (line) layer'),
+                [QgsProcessing.TypeVectorLine]
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFeatureSink(
+                self.OUTPUT,
+                self.tr('Climb layer')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.TOTALCLIMB,
+                self.tr('Total climb')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.TOTALDESCENT,
+                self.tr('Total descent')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.MINELEVATION,
+                self.tr('Minimum elevation')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.MAXELEVATION,
+                self.tr('Maximum elevation')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        fcount = source.featureCount()
+
+        hasZ = QgsWkbTypes.hasZ(source.wkbType())
+
+        if not hasZ:
+            feedback.reportError(self.tr('The layer does not have Z values. If you have a DEM, use the Drape algorithm to extract Z values.'))
+            return
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        minelevindex = -1
+        maxelevindex = -1
+        fieldnumber = 0
+
+        # Skip fields with names that are equal to the generated ones
+        for field in source.fields():
```

Comment:
I'm unsure about this block -- elsewhere we create two lists of fields (existing and new), and then use QgsProcessingUtils::combineFields to ensure field have unique names when there's a clash. I think that same approach should be used here.

Reply:
The problem with using combineFields is that the names of the climb, descent, ... fields will not be know in the output (could be "climb", "climb_1", "climb_23", ...).  This will make it difficult to use this algorithm in a chain of operations.  So I think the original solution is better in this case.

Reply:
@nyalldawson, I have now suggested a solution where combineFields is still used, but with the new fields as the first parameter and the input fields as the second parameter.  Is that a solution you can accept?